wherescape/connectors/gitlab/gitlab_wrapper.py

```diff
@@ -0,0 +1,198 @@
+        pagination = f"per_page={page_variables['per_page']}&page={int(page_variables['current_page'])+1}"
+        return f"{self.base_url}/{resource_api}?order_by={order_by}&sort=asc&simple={simple}&{pagination}{updated_since}"
+
+    def paginate_through_resource(
+        self,
+        resource_api,
+        keys_to_keep,
+        since=None,
+        per_page=50,
+        simple="false",
+        order_by="id",
+    ):
+        """Paginate through resources
+        Since the Gitlab API has pagination, this helper function will paginate through the resource API.
+        It will do that until all responses are collected.
+        It cleans the response immediately and turns it into a tuple.
+        It does expect a list with objects in the response of the API.
+
+        Parameters:
+        resource_api (string): The location of the resource requested
+        keys_to_keep (list): List of keys returned by the API you want to keep
+        per_page (int): How many results per page you would like to get
+        simple (boolean): If the response of Gitlab should be simplified this needs to be set on True
+        since (string): ISO formatted datetime string to indicate since which date you want values back
+
+        Returns:
+        List of tuples with the values from the request
+
+        """
+        total_pages = 1
+        current_page = 0
+
+        all_resources = []
+
+        while current_page < total_pages:
+            page_variables = {"per_page": per_page, "current_page": current_page}
+            url = self.format_url(resource_api, page_variables, simple, order_by, since)
+
+            response = self.make_request(url, "GET")
+
+            if response.status_code == 403:
+                logging.info(
+                    f"{url}\n Forbidden resource. If you need this resource, please check the user's rights"
+                )
+                current_page = current_page + 1
+                continue
+
+            response.raise_for_status()
+
+            json_response = response.json()
+
+            for resource_object in json_response:
+                cleaned_json = filter_dict(flatten_json(resource_object), keys_to_keep)
+                final_json = fill_out_empty_keys(cleaned_json, keys_to_keep)
+                all_resources.append(list(final_json.values()))
+
+            try:
+                total_pages = response.headers["X-Total-Pages"]
+                current_page = response.headers["X-Page"]
+            except:
+                current_page = current_page + 1
+
+        return all_resources
+
+    def get_projects(self):
+        return self.projects
+
+    def get_projects_from_api(self):
+        keys_to_keep = COLUMN_NAMES_AND_DATA_TYPES["projects"].keys()
+        resource_api = "projects"
+
+        all_projects = self.paginate_through_resource(
+            resource_api, keys_to_keep, simple="true"
+        )
+        return all_projects
+
+    def get_release_tags(self):
```

Comment:
I would expect this to be called`get_release_tags_from_api`

Reply:
also for other functions

Reply:
Changed the names around! Made the get_projects to get_projects_loaded to indicate it will get prefetched data.